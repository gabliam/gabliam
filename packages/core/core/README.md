[![Build Status][build-image]][build-url]
[![NPM version][npm-image]][npm-url]


# Gabliam core

Gabliam is on top of inversify.
Gabliam help you for register all class in inversify.
Gabilliam is modular with plugins.


# Gabliam phase

## Build phase

1. Load all config file
2. Load all file in scanPath of main
	- All decorators on class add metadata and can register this class
	- if an class has @scanPath, add this folder in queue for load
	- load all plugin
3. Bind classes in DI (inversify)
	- call all plugin.bind
4. load all config classes (order: CoreConfig => PluginConfig => config)
	- call all plugin.config for all instance of config class
5. Build
	- call all plugin.build

## Start phase
/!\ Build phase must be passed

1. call all plugin.start

## Stop phase
/!\ Build and start phases must be passed

1. call all plugin.start

## Detroy phase
/!\ Stop phase must be passed

1. call all plugin.detroy

# Samples
Go to [Samples]



[Samples]: <https://github.com/gabliam/gabliam/tree/master/examples>

# License

  MIT

[build-image]: https://img.shields.io/travis/gabliam/gabliam/master.svg?style=flat-square
[build-url]: https://travis-ci.org/gabliam/gabliam
[npm-image]: https://img.shields.io/npm/v/@gabliam/core.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@gabliam/gabliam/core
