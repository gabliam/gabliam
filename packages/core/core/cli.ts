#!/usr/bin/env node
import 'reflect-metadata';
import yargs from 'yargs';
import { StartCommand } from './src/commands/start-command';

yargs
  .usage('Usage $0 <command> [options]')
  .command(new StartCommand())
  .recommendCommands()
  .demandCommand(1)
  .strict()
  .alias('v', 'version')
  .help('h')
  .alias('h', 'help').argv;

require('yargonaut')
  .style('blue')
  .style('yellow', 'required')
  .helpStyle('green')
  .errorsStyle('red');
