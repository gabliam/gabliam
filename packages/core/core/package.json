{
  "name": "@gabliam/core",
  "version": "0.0.0-PLACEHOLDER",
  "description": "Gabliam is on top of inversify",
  "main": "index.js",
  "typings": "index.d.ts",
  "module": "index.js",
  "scripts": {
    "build": "rimraf lib && tsc -p tsconfig.build.json"
  },
  "files": [
    "lib",
    "src",
    "index.js",
    "index.d.ts"
  ],
  "dependencies": {
    "@gabliam/expression": "0.0.0-PLACEHOLDER",
    "@types/joi": "^14.0.1",
    "caller": "^1.0.1",
    "co": "^4.6.0",
    "debug": "^4.1.1",
    "glob": "^7.1.3",
    "graph-data-structure": "^1.8.0",
    "inversify": "^5.0.1",
    "inversify-logger-middleware": "^3.0.0",
    "joi": "^14.3.1",
    "js-yaml": "^3.12.1",
    "lodash": "^4.17.11",
    "shortstop": "^1.0.3",
    "shortstop-handlers": "^1.0.1",
    "tslib": "^1.9.3"
  },
  "devDependencies": {
    "@types/debug": "^4.1.3",
    "@types/fs-extra": "^5.0.4",
    "@types/glob": "^7.1.1",
    "@types/js-yaml": "^3.11.4",
    "@types/lodash": "^4.14.123",
    "fs-extra": "^7.0.1",
    "properties": "^1.2.1",
    "rxjs": "^6.3.3",
    "toml": "^2.3.3"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/gabliam/gabliam.git"
  },
  "author": "David Touzet <dtouzet@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/gabliam/gabliam/issues"
  },
  "homepage": "https://github.com/gabliam/gabliam/tree/master/packages/core#readme",
  "engines": {
    "node": ">=8.9.0"
  },
  "gitHead": "ec328fa0b62f474d470f74a2fc9637838f54091e"
}
