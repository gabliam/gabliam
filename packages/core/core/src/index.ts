import 'reflect-metadata';

export {
  decorate,
  inject,
  injectable,
  interfaces as inversifyInterfaces,
  multiInject,
  named,
  optional,
  postConstruct,
  tagged,
  targetName,
  unmanaged,
} from 'inversify';
export * from './common';
export {
  APP_CONFIG,
  CORE_CONFIG,
  INJECT_CONTAINER_KEY,
  VALUE_EXTRACTOR,
} from './constants';
export * from './container';
export * from './decorator';
export * from './errors';
export * from './gabliam';
export * from './gabliam-options';
export * from './gabliam-utils';
export * from './interfaces';
export * from './joi';
export { configResolver, FileLoader, Loader, Resolver } from './loaders';
export * from './metadatas';
export * from './reflection';
export * from './registry';
export * from './value-extractor';
