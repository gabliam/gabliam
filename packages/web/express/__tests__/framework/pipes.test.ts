import { Config, Container, Service } from '@gabliam/core';
import {
  Controller,
  Get,
  Pipe,
  RequestParam,
  UsePipes,
  WebConfig,
  WebConfiguration,
} from '@gabliam/web-core';
import { WebPluginTest } from '@gabliam/web-core/lib/testing';
import * as sinon from 'sinon';
import ExpressPlugin from '../../src';

let appTest: WebPluginTest;

beforeEach(async () => {
  appTest = new WebPluginTest([ExpressPlugin]);
});

afterEach(async () => {
  await appTest.destroy();
});

describe('pipes', () => {
  @Service()
  class A implements Pipe {
    transform(value: any) {
      return (value += 'a');
    }
  }

  @Service()
  class B implements Pipe {
    transform(value: any) {
      return (value += 'b');
    }
  }

  @Service()
  class C implements Pipe {
    transform(value: any) {
      return (value += 'c');
    }
  }

  const spyA = sinon.spy(A.prototype, 'transform');
  const spyB = sinon.spy(B.prototype, 'transform');
  const spyC = sinon.spy(C.prototype, 'transform');

  beforeEach(() => {
    spyA.resetHistory();
    spyB.resetHistory();
    spyC.resetHistory();
  });

  test('should call method-level Pipe correctly', async () => {
    @Controller('/')
    class TestController {
      @UsePipes(A, B, C)
      @Get('/:id')
      public getTest(@RequestParam('id') id: string) {
        return id;
      }
    }
    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/lol')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
  });

  test('should call controller-level Pipe correctly', async () => {
    @UsePipes(A, B, C)
    @Controller('/')
    class TestController {
      @Get('/:id')
      public getTest(@RequestParam('id') id: string) {
        return id;
      }
    }
    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/lol')
      .expect(200);

    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
  });

  test('`should call server-level pipe correctly`', async () => {
    @Controller('/')
    class TestController {
      @Get('/:id')
      public getTest(@RequestParam('id') id: string) {
        return id;
      }
    }

    @Config()
    class ServerConfig {
      @WebConfig()
      serverConfig(app: any, container: Container) {
        const webConfiguration = container.get(WebConfiguration);
        webConfiguration.useGlobalPipes(A, B, C);
      }
    }

    appTest.addClass(TestController);
    appTest.addClass(ServerConfig);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/lol')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
  });

  test('should call all pipe in correct order', async () => {
    @UsePipes(B)
    @Controller('/')
    class TestController {
      @UsePipes(C)
      @Get('/:id')
      public getTest(@RequestParam('id') id: string) {
        return id;
      }
    }

    @Config()
    class ServerConfig {
      @WebConfig()
      serverConfig(app: any, container: Container) {
        const webConfiguration = container.get(WebConfiguration);
        webConfiguration.useGlobalPipes(A);
      }
    }

    appTest.addClass(TestController);
    appTest.addClass(ServerConfig);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/lol')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
  });

  test('should resolve controller-level pipe', async () => {
    const symbolId = Symbol('spyA');
    const strId = 'spyB';

    @UsePipes(symbolId, strId)
    @Controller({
      path: '/',
    })
    class TestController {
      @Get('/:id')
      public getTest(@RequestParam('id') id: string) {
        return id;
      }
    }

    appTest.addClass(TestController);
    appTest.gab.container
      .bind(symbolId)
      .to(A)
      .inSingletonScope();
    appTest.gab.container
      .bind(strId)
      .to(B)
      .inSingletonScope();
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/lol')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
  });

  test('should resolve method-level pipe', async () => {
    const symbolId = Symbol('spyA');
    const strId = 'spyB';

    @Controller('/')
    class TestController {
      @UsePipes(symbolId, strId)
      @Get('/:id')
      public getTest(@RequestParam('id') id: string) {
        return id;
      }
    }

    appTest.addClass(TestController);
    appTest.gab.container
      .bind(symbolId)
      .to(A)
      .inSingletonScope();
    appTest.gab.container
      .bind(strId)
      .to(B)
      .inSingletonScope();
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/lol')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
  });

  test('should compose controller- and method-level pipe', async () => {
    const symbolId = Symbol('spyA');
    const strId = 'spyB';

    @UsePipes(symbolId)
    @Controller({
      path: '/',
    })
    class TestController {
      @UsePipes(strId)
      @Get('/:id')
      public getTest(@RequestParam('id') id: string) {
        return id;
      }
    }

    appTest.addClass(TestController);
    appTest.gab.container
      .bind(symbolId)
      .to(A)
      .inSingletonScope();
    appTest.gab.container
      .bind(strId)
      .to(B)
      .inSingletonScope();
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/lol')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
  });

  test('should compose controller- and method-level pipe and can use an instance', async () => {
    class D implements Pipe {
      constructor(private prefix: string) {}
      transform(value: any) {
        return (value += this.prefix + 'c');
      }
    }

    @UsePipes(A)
    @Controller({
      path: '/',
    })
    class TestController {
      @UsePipes(B, C, new D('||'))
      @Get('/:id')
      public getTest(@RequestParam('id') id: string) {
        return id;
      }
    }

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/lol')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
  });
});
