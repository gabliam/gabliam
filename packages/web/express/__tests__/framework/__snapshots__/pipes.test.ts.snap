// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pipes \`should call server-level pipe correctly\` 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "lolabc",
}
`;

exports[`pipes should call all pipe in correct order 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "lolabc",
}
`;

exports[`pipes should call controller-level Pipe correctly 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "lolabc",
}
`;

exports[`pipes should call method-level Pipe correctly 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "lolabc",
}
`;

exports[`pipes should compose controller- and method-level pipe 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "lolab",
}
`;

exports[`pipes should compose controller- and method-level pipe and can use an instance 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "9",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "lolabc||c",
}
`;

exports[`pipes should resolve controller-level pipe 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "lolab",
}
`;

exports[`pipes should resolve method-level pipe 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "lolab",
}
`;
